### PR TITLE
Allow calculation on subset of data

### PR DESCRIPTION
**Pull request by roya0045**
## Description
supersede #9814 

@nyalldawson sorry for the issues. This should be it, there is also an implementation in qgsvectorlayer but I'll leave it in my bigger PR for now.

I am aware that tests would be good but they will have to wait a little.

### CODE REVIEW

File: python/core/auto_generated/qgsaggregatecalculator.sip.in
Comment (m-kuhn) [inline python/core/auto_generated/qgsaggregatecalculator.sip.in]: How about a QgsFeatureRequest instead of the ids here?

Comment (roya0045) [reply python/core/auto_generated/qgsaggregatecalculator.sip.in]: No, the bigger PR is dependent on this. If needed someone could overload the function but in order to perform calculations on every feature associated with a given symbol I need to know which feature is associated with which symbol. To have access to this this list of symbol is stored in the featurecounter and is provided by the symbol expression.

Yes this is the part that causes segfault but I don't see any workaround to properly link the symbols to their corresponding features and passing it down to the calculator.

The feature ids are passed to the featurerequest used in the aggregator.

Comment (roya0045) [reply python/core/auto_generated/qgsaggregatecalculator.sip.in]: @m-kuhn if we wanted to provide a feature request the optimal thing would be to leave the head of this function as a generator for a QgsFeatureRequest and move the code of the core to another function that would directly use the qgsFeatureRequest, similar to what Mr. Dawson did where he just split the function into its internal part and made front-facing code that would allow users to provide their inputs for iteration.

But this is out of the scope of this PR.

File: tests/src/python/test_qgsaggregatecalculator.py
Comment (nyalldawson) [inline tests/src/python/test_qgsaggregatecalculator.py]: Can you also add a test with an empty list? (I.e. should be no features used)

File: python/core/auto_generated/qgsvectordataprovider.sip.in
Comment (nyalldawson) [inline python/core/auto_generated/qgsvectordataprovider.sip.in]: "Pointer to a list of feature IDs to use in the calculation. Set to NULLPTR to disable any ID based filtering. An empty list indicates that no features will be considered."

(Make sure NULLPTR is all capitals, it's a special keyword)

File: tests/src/python/test_qgsaggregatecalculator.py
Comment (roya0045) [reply tests/src/python/test_qgsaggregatecalculator.py]: done

File: python/core/auto_generated/qgsvectordataprovider.sip.in
Comment (roya0045) [reply python/core/auto_generated/qgsvectordataprovider.sip.in]: Done

**Comment by roya0045** [comment]
@m-kuhn are you ok with the current proposal and the justifications that I provided?

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Looks good to me!

File: python/core/auto_generated/qgsaggregatecalculator.sip.in
Comment (m-kuhn) [reply python/core/auto_generated/qgsaggregatecalculator.sip.in]: Internally, the ids are fed into a QgsFeatureRequest. I think it should be trivial to start from a request that is provided as a parameter instead of a new one there. Did I overlook something?

Comment (roya0045) [reply python/core/auto_generated/qgsaggregatecalculator.sip.in]: Yes, a feature request is created already in the original implementation, all this does is use the proper parameter of the featurerequest to filter out unwanted features, this does not add a new featurerequest. Also inputing directly a feature request might be less api/user friendly as this is already handled internally in the function. We'll go over that once we get to talk.

File: src/core/qgsaggregatecalculator.cpp
Comment (m-kuhn) [inline src/core/qgsaggregatecalculator.cpp]:
```suggestion
    const QgsFeatureRequest request = QgsFeatureRequest() ) const
```

**Comment by roya0045** [comment]
@nyalldawson I have adapted the code to Mr.Kuhn's suggestion. So instead of using a list of features to filter, a preset featurequest will be used instead. Is that ok to you?

File: src/core/qgsaggregatecalculator.cpp
Comment (roya0045) [reply src/core/qgsaggregatecalculator.cpp]: changes have been commited

**Comment by roya0045** [comment]
Tests are behaving weirdly, test failed due to a 503 docker error. (also travis has issues with normal tests now but not code layout test.) @nyalldawson @m-kuhn

**Comment by roya0045** [comment]
@nyalldawson @m-kuhn I'm not even sure what is causing the code layout to fail when the 10 tests have passed.

**Comment by m-kuhn** [comment]
I would strongly recommend to revert filter stacking in this pull request.
Because it needs some discussion of concepts and approaches, this will block the other functionality for quite some time.

Meanwhile, you can easily use 

    '"attrib" = \'some filter criteria\' AND $id in ({ids})'.format(ids=','.join(id_list))

**Comment by nyalldawson** [comment]
I second what @m-kuhn says -- please revert the request changes, and I'll re-review.

**Comment by roya0045** [comment]
@nyalldawson @m-kuhn Agreed, I just wanted to make something quick in order to have a semblance of idea to what such a thing might look like.

PR should be OK for review unless I made a typo.

**Comment by roya0045** [comment]
> I would strongly recommend to revert filter stacking in this pull request.
> Because it needs some discussion of concepts and approaches, this will block the other functionality for quite some time.
> 
> Meanwhile, you can use
> 
> ```
> '"attrib" = \'some filter criteria\' AND $id in ({ids})'.format(ids=','.join(id_list))
> ```

Yes but since I'm using a qgsrequest as the input to filter fids, this would mean that any expression will override this fids, and this also mean that one cannot change easily this fact by adding the list of filtered fids to the expression, when it comes to evaluation at symbol level.

**Comment by roya0045** [comment]
@m-kuhn @nyalldawson regarding using fids and filters, would it be possible to make some changes that would use the fids provided instead of checking every fid of a layer to see if it's in the list. This would greatly accelerate things and would remove one check but I'm not sure if such and implementation would be easy to do.

File: src/core/qgsaggregatecalculator.cpp
Comment (nyalldawson) [inline src/core/qgsaggregatecalculator.cpp]: @m-kuhn What's your thoughts on this? The danger here is that if someone specifies a feature request with an existing filter expression or filter fids etc, it'll get silently overridden by this.

I'm 50/50 on it. I don't think it's a big issue so long as the dox for the the request parameter and for setFilter explicitly state this situation using "\warn" tags

File: src/core/qgsvectordataprovider.h
Comment (nyalldawson) [inline src/core/qgsvectordataprovider.h]:
```suggestion
     * \param request optional feature request with feature filtering criteria, if not provided all features will be used
```

File: src/core/qgsaggregatecalculator.cpp
Comment (roya0045) [reply src/core/qgsaggregatecalculator.cpp]: Yeah I didn't want to do this but Mr. Kuhn was proving insistent with this instead of the feature Ids. But as stated in the emails if someone sets and expression it overrides the filter lists. This is why I wanted to address this issue.

Initially using the filter list was simple but since it was lower than the expression in  the order this meant that the expression was overridden but the feature list.

Once again in both cases we are at a loss. The tests that I did with my previous proposal worked somehwat fine but could lead to segfault if stacking was allowed and no expression was provided.

I don't mind using either to be honest.

File: src/core/qgsvectordataprovider.h
Comment (roya0045) [reply src/core/qgsvectordataprovider.h]: done

File: src/core/qgsaggregatecalculator.cpp
Comment (m-kuhn) [reply src/core/qgsaggregatecalculator.cpp]: Oh, I wasn't aware there already is a filter possibility (via a mFilterExpression member) on the aggregate calculator.

In this case, shouldn't the feature ids just be inserted in there?
Or the list of filter ids be injected via setParameters too?

Comment (roya0045) [reply src/core/qgsaggregatecalculator.cpp]: > Oh, I wasn't aware there already is a filter possibility (via a mFilterExpression member) on the aggregate calculator.
> 
> In this case, shouldn't the feature ids just be inserted in there?
> Or the list of filter ids be injected via setParameters too?

...so I'll just revert everything to its original state. :confused:

Comment (roya0045) [reply src/core/qgsaggregatecalculator.cpp]: Ignore comment about reseting,

As far as the expression part yes this could be done but as I stated previously currently the expressions and variables are more QVariant and support for QVariantList is limited. This implies recasting everything for storage and to get it back as a QgsFeatureIds, this also mean that I would need to find a way to retrieve that list as a variable and currently support for list retrieval is not implemented from what I tested, but I may be wrong.

This is why I commented out the use of Symbol_features (can't recall the name) in my other PR.

Comment (roya0045) [reply src/core/qgsaggregatecalculator.cpp]: @m-kuhn So should I revert to filter ids or leave it as a custom request?

**Comment by roya0045** [comment]
With more fidling around using the QgsFeatureRequest as the input might be better since for some aggregation there are no easy way to specify the expression filter, with the request you can specify both, which might be better.

This would be an alternative to set params in the API.

@nyalldawson @m-kuhn

**Comment by roya0045** [comment]
@nyalldawson @m-kuhn I'll be adding a warning until my other PR is merged, can we get this merged?

As stated exposing the feature request allows to better control the filtering mechanisms and implement the proper triggers in the iterator to allow the use of multiple filters.

**Comment by roya0045** [comment]
I have fixed the dox. @m-kuhn Are you ok with everything? Fiddling with parameter setting might be more cannon. Do you want me to alter setparameter and trigger stacking in the function call to aggregate?

**Comment by roya0045** [comment]
@nyalldawson @m-kuhn I have reverted the input for vectorlayer and vectorlayerdataprovider, I felt that the request was out of place and that the fids pointer was better.

This means that only the matter of the aggregate calculator is left and that the handling of that will be done in the vectorlayer and in the expressionfunction(other PR) when needed.

I will merge this with my FIDS pr for consistency.

**Comment by roya0045** [comment]
Wait, I just realized that exposing the request allows to use a base field, + any types of filters, rather than input filtering in the field parameter and using where conditions, isn't that an improvement? (I'll revert things in that case) @m-kuhn @nachouve

**Comment by roya0045** [comment]
@m-kuhn I did something that is more consistent with the API, but this provide no way to control the injection order in the request as it is defined and the expression will always be evaluated instead of fids if stacking is not allowed.

This should be more API consistent.

**Comment by roya0045** [comment]
@nyalldawson Are you still ok with the changes?

**Comment by roya0045** [comment]
@nyalldawson Just a quick ping to see if you are still ok with the API and docs. I'm not too sure if the documentation is easy to understand at the moment or if rewording it might make things clearer.

**Comment by roya0045** [comment]
@nyalldawson I'm not sure if Matthias is awaiting your reply about the approval in light of the changes that I made. Hence the weekly ping to ensure that approval is maintained And to prevent stale status.

**Comment by roya0045** [comment]
Not stale, awaiting validation from @nyalldawson .

File: src/core/qgsaggregatecalculator.h
Comment (nyalldawson) [inline src/core/qgsaggregatecalculator.h]: Can you remove this until after #9915 is finalised please?

**Review by nyalldawson — COMMENTED** [review]
I'm happy with the current iteration, if you can just remove the references to stacking (currently unused and dependant on #9915)

File: src/core/qgsaggregatecalculator.cpp
Comment (m-kuhn) [inline src/core/qgsaggregatecalculator.cpp]: const ref
```suggestion
void QgsAggregateCalculator::setFidsFilter( const QgsFeatureIds &fids )
```

File: src/core/qgsaggregatecalculator.h
Comment (roya0045) [reply src/core/qgsaggregatecalculator.h]: Done, thanks again for the review!

File: src/core/qgsaggregatecalculator.cpp
Comment (roya0045) [reply src/core/qgsaggregatecalculator.cpp]: An oversight on my part, should be fixed now, thank you for the review!

**Comment by roya0045** [comment]
@nyalldawson Thanks for the merge!

**Pull request merged**
